File: dnn-providers/hip-kernel-provider/rocke/platform/Python/rocke/helpers/streamk.py
# Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
# SPDX-License-Identifier: MIT

"""StreamK tile partitioner helpers (CK Tile parity).

CK Tile's :class:`StreamKTilePartitioner` decomposes a GEMM workload
into ``(m_tile, n_tile, k_iter)`` triples and decides how each
partial-K accumulation feeds the output. Two reduction strategies are
exposed:

* ``ReductionStrategy.Atomic`` -- each CTA atomically adds its
 partial K-sum into a shared f32 workspace at the output's
 ``(m_tile, n_tile)`` position. Simpler; relies on
 ``global_atomic_add_f32`` (gfx940+). Requires a finalisation pass
 to convert f32 -> output dtype.

* ``ReductionStrategy.Reduction`` -- the CTAs that contribute to the
 same ``(m_tile, n_tile)`` cooperate through a tile-major workspace
 + a flag table; the *last* contributor performs the reduction +
 finalisation in-kernel. More complex but avoids the second launch
 and is the path CK Tile defaults to.

This module ships the partitioner math (a pure-Python view) and the
IR-side glue that
``instances/streamk_gemm.py`` uses. The kernel-shape primitives stay
in :mod:`rocke.helpers.atoms` / :mod:`rocke.helpers.loads` /
:mod:`rocke.helpers.epilogues`; this file is purely about
*partitioning* the work, not about running the per-tile GEMM.

What we ship today:

* :class:`StreamKReductionStrategy` -- enum of supported strategies.
* :class:`StreamKPartition` -- the decoded ``(m_tile, n_tile, k_iter,
 is_first, is_last)`` for a given linear partition id.
* :func:`compute_streamk_grid_size` -- worst-case CTA count for a
 spec.
* :func:`emit_streamk_decode` -- IR-side decode of a linear partition
 id into the SSA ``(m_tile, n_tile, k_iter, is_first, is_last)``
 bundle the kernel needs.

v1 implements the ``Atomic`` strategy end-to-end and the
``Reduction`` strategy's *decode* surface (so the spec layer can
emit the right kernel shape); the per-tile reduction pass that the
``Reduction`` strategy needs lands with the StreamK GEMM kernel
itself in .
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import NamedTuple

from ..core.ir import IRBuilder, Value


__all__ = [
    "StreamKPartition",
    "StreamKReductionStrategy",
    "compute_streamk_grid_size",
    "emit_streamk_decode",
    "emit_streamk_partial_load_accumulate",
    "emit_streamk_partial_store",
    "streamk_num_macro_tiles",
]


class StreamKReductionStrategy(Enum):
    """Per CK Tile naming."""

    Atomic = "atomic"  # global_atomic_add_f32 into workspace
    Reduction = "reduction"  # cooperative + flag-table reduction


@dataclass(frozen=True)
class StreamKPartition:
    """The compile-time-fixed shape inputs to the StreamK partitioner.

    The actual ``(m_tile, n_tile, k_iter, is_first, is_last)``
    decoding happens at runtime via :func:`emit_streamk_decode`;
    this dataclass just collects the compile-time constants the
    decode helper needs.
    """

    m_tiles: int  # ceil(M / tile_m); the number of M tiles
    n_tiles: int  # ceil(N / tile_n)
    k_iters: int  # K / tile_k

    @property
    def num_macro_tiles(self) -> int:
        """``m_tiles * n_tiles * k_iters`` -- the total chunk count."""
        return self.m_tiles * self.n_tiles * self.k_iters

    @property
    def k_iters_per_output_tile(self) -> int:
        """``k_iters`` -- the number of CTAs that touch each
        ``(m, n)`` output tile in the Atomic strategy.
        """
        return self.k_iters


class _DecodedTile(NamedTuple):
    """The SSA bundle returned by :func:`emit_streamk_decode`."""

    m_tile: Value  # i32
    n_tile: Value  # i32
    k_iter: Value  # i32 in [0, k_iters)
    is_first: Value  # i1: k_iter == 0
    is_last: Value  # i1: k_iter == k_iters - 1


def streamk_num_macro_tiles(spec: StreamKPartition) -> int:
    """Plain Python view of :attr:`StreamKPartition.num_macro_tiles`."""
    return spec.num_macro_tiles


def compute_streamk_grid_size(
    spec: StreamKPartition,
    *,
    num_cus: int = 304,
    blocks_per_cu: int = 1,
) -> int:
    """Recommended persistent-launch grid size for ``spec``.

    Cap = ``min(num_macro_tiles, num_cus * blocks_per_cu)``. The
    persistent kernel re-fetches macro tiles via atomic_add until
    the global counter exhausts; the launch grid stays small and
    constant so launch overhead is one-shot.

    Default ``num_cus = 304`` matches MI300X (the canonical CK Tile
    deployment target). Override for MI355X (304) or other parts.
    """
    if spec.num_macro_tiles <= 0:
        raise ValueError("spec has zero macro tiles")
    return min(spec.num_macro_tiles, num_cus * blocks_per_cu)


def emit_streamk_decode(
    b: IRBuilder, linear_id: Value, spec: StreamKPartition
) -> _DecodedTile:
    """Decode a linear macro-tile id into ``(m_tile, n_tile, k_iter,
    is_first, is_last)``.

    Layout (matches CK Tile's "K-major within a (m, n) tile" walk):

    .. code-block:: text

    k_iter = linear_id % k_iters
    nn = linear_id // k_iters
    n_tile = nn % n_tiles
    m_tile = nn // n_tiles

    The ``is_first`` / ``is_last`` predicates let the kernel decide
    whether it owns the *seed* of the output (must clear) or the
    *finalize* (must write the converted output dtype to ``C``); the
    middle iterations atomic-add their partial into the workspace.
    """
    c_k_iters = b.const_i32(spec.k_iters)
    c_n_tiles = b.const_i32(spec.n_tiles)

    k_iter = b.mod(linear_id, c_k_iters)
    nn = b.div(linear_id, c_k_iters)
    n_tile = b.mod(nn, c_n_tiles)
    m_tile = b.div(nn, c_n_tiles)

    is_first = b.cmp_eq(k_iter, b.const_i32(0))
    is_last = b.cmp_eq(k_iter, b.const_i32(spec.k_iters - 1))
    return _DecodedTile(
        m_tile=m_tile,
        n_tile=n_tile,
        k_iter=k_iter,
        is_first=is_first,
        is_last=is_last,
    )


def emit_streamk_partial_store(
    b: IRBuilder,
    *,
    workspace: Value,
    flag_table: Value,
    workspace_off: Value,
    flag_off: Value,
    value: Value,
    is_first: Value,
) -> None:
    """Cooperative ``Reduction`` strategy partial-store + signal.

    P37: each contributing CTA stores its partial K-sum at the
    ``(m_tile, n_tile, k_iter)`` slot in the workspace and bumps the
    flag-table counter for ``(m_tile, n_tile)`` via ``atomic_add(+1)``.
    The first contributor's store is unconditionally a fresh write
    (``is_first=True``); later contributors atomic-add into the
    workspace too so the order doesn't matter.

    Reference: CK Tile ``streamk_common.hpp::SignalStorePartialDone``
    + ``streamk_gemm_kernel.hpp:448-504`` (Tree-reduction).
    """
    # Initial store: lane-0-only write (the workspace is per-lane sized
    # already in the caller's view) — the consumer in
    # :func:`emit_streamk_partial_load_accumulate` reads each slot once.
    with b.scf_if(is_first):
        b.global_store(workspace, workspace_off, value, align=4)
    # Non-first contributors: atomic add into the same f32 slot (the
    # ``is_last`` reducer reads the converged value).
    with b.scf_if(b.lnot(is_first)):
        b.global_atomic_add(workspace, workspace_off, value)
    # Bump the flag counter so the last-finishing CTA can detect "I'm
    # the reducer" via flag == k_iters_per_tile.
    b.global_atomic_add(flag_table, flag_off, b.const_i32(1))


def emit_streamk_partial_load_accumulate(
    b: IRBuilder,
    *,
    workspace: Value,
    workspace_off: Value,
    is_last: Value,
) -> Value:
    """Read the converged f32 partial sum at ``workspace_off`` once
    every contributor has signalled (``is_last == True`` is the
    caller's responsibility).

    The flag-table-based wait protocol is a busy-loop on
    ``atomic_load(flag_table[tile_id]) == k_iters_per_tile``; we expose
    it as a separate helper because the busy-loop shape differs
    between Linear / Tree reduction. For now the Atomic strategy is
    the canonical path; this helper plus
    :func:`emit_streamk_partial_store` give callers the surface
    needed to opt into the Reduction strategy.

    Reference: CK Tile ``streamk_common.hpp:34-73``
    (``WaitStorePartialDone``).
    """
    return b.global_load_f32(workspace, workspace_off)

File: dnn-providers/hip-kernel-provider/rocke/platform/Cpp/helpers/streamk.cpp
// Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
// SPDX-License-Identifier: MIT
/*
 * helper_rocke.helpers.streamk.c -- C99 port of rocke.helpers.streamk.
 *
 * Ports the four partitioner symbols:
 *   StreamKReductionStrategy, StreamKPartition, compute_streamk_grid_size,
 *   emit_streamk_decode.
 *
 * compute_streamk_grid_size is pure-int; emit_streamk_decode's builder-call
 * sequence is byte-identical to the Python so the emitted IR op stream matches
 * exactly.
 */
#include "rocke/helper_rocke.helpers.streamk.h"

#include <stddef.h> /* NULL */

#include "rocke/ir_internal.h" /* rocke_i_set_err for Python-ValueError parity */

/* ------------------------------------------------------------------------
 * StreamKReductionStrategy enum value strings (CK Tile naming).
 * ------------------------------------------------------------------------ */
const char* rocke_streamk_reduction_strategy_value(rocke_streamk_reduction_strategy_t s)
{
    switch(s)
    {
    case ROCKE_STREAMK_REDUCTION_ATOMIC:
        return "atomic";
    case ROCKE_STREAMK_REDUCTION_REDUCTION:
        return "reduction";
    default:
        return NULL;
    }
}

/* ------------------------------------------------------------------------
 * StreamKPartition properties.
 *
 *   @property num_macro_tiles: m_tiles * n_tiles * k_iters
 *   @property k_iters_per_output_tile: k_iters
 * ------------------------------------------------------------------------ */
int rocke_streamk_partition_num_macro_tiles(const rocke_streamk_partition_t* spec)
{
    return spec->m_tiles * spec->n_tiles * spec->k_iters;
}

int rocke_streamk_partition_k_iters_per_output_tile(const rocke_streamk_partition_t* spec)
{
    return spec->k_iters;
}

/* Module-level streamk_num_macro_tiles(spec): plain Python view. */
int rocke_streamk_num_macro_tiles(const rocke_streamk_partition_t* spec)
{
    return rocke_streamk_partition_num_macro_tiles(spec);
}

/* ------------------------------------------------------------------------
 * compute_streamk_grid_size
 *
 *   if spec.num_macro_tiles <= 0:
 *       raise ValueError("spec has zero macro tiles")
 *   return min(spec.num_macro_tiles, num_cus * blocks_per_cu)
 * ------------------------------------------------------------------------ */
int rocke_compute_streamk_grid_size(const rocke_streamk_partition_t* spec,
                                    int num_cus,
                                    int blocks_per_cu,
                                    rocke_status_t* out_status)
{
    int num_macro_tiles;
    int cap;

    num_macro_tiles = rocke_streamk_partition_num_macro_tiles(spec);
    if(num_macro_tiles <= 0)
    {
        if(out_status != NULL)
        {
            *out_status = ROCKE_ERR_VALUE;
        }
        return -1; /* Python: raise ValueError("spec has zero macro tiles") */
    }

    cap = num_cus * blocks_per_cu;
    if(out_status != NULL)
    {
        *out_status = ROCKE_OK;
    }
    return (num_macro_tiles < cap) ? num_macro_tiles : cap;
}

/* ------------------------------------------------------------------------
 * emit_streamk_decode
 *
 *   c_k_iters = b.const_i32(spec.k_iters)
 *   c_n_tiles = b.const_i32(spec.n_tiles)
 *   k_iter    = b.mod(linear_id, c_k_iters)
 *   nn        = b.div(linear_id, c_k_iters)
 *   n_tile    = b.mod(nn, c_n_tiles)
 *   m_tile    = b.div(nn, c_n_tiles)
 *   is_first  = b.cmp_eq(k_iter, b.const_i32(0))
 *   is_last   = b.cmp_eq(k_iter, b.const_i32(spec.k_iters - 1))
 *   return (m_tile, n_tile, k_iter, is_first, is_last)
 *
 * The Python evaluates b.const_i32(0) and b.const_i32(spec.k_iters - 1) as
 * arguments inside the cmp_eq calls; C's argument evaluation order is
 * unspecified, so pin the const-then-cmp order with explicit temporaries.
 * ------------------------------------------------------------------------ */
rocke_streamk_decoded_tile_t rocke_emit_streamk_decode(rocke_ir_builder_t* b,
                                                       rocke_value_t* linear_id,
                                                       const rocke_streamk_partition_t* spec)
{
    rocke_streamk_decoded_tile_t res;
    rocke_value_t* c_k_iters;
    rocke_value_t* c_n_tiles;
    rocke_value_t* nn;
    rocke_value_t* c_zero;
    rocke_value_t* c_last;

    res.m_tile = NULL;
    res.n_tile = NULL;
    res.k_iter = NULL;
    res.is_first = NULL;
    res.is_last = NULL;

    c_k_iters = rocke_b_const_i32(b, (int64_t)spec->k_iters);
    c_n_tiles = rocke_b_const_i32(b, (int64_t)spec->n_tiles);

    res.k_iter = rocke_b_mod(b, linear_id, c_k_iters);
    nn = rocke_b_div(b, linear_id, c_k_iters);
    res.n_tile = rocke_b_mod(b, nn, c_n_tiles);
    res.m_tile = rocke_b_div(b, nn, c_n_tiles);

    c_zero = rocke_b_const_i32(b, 0);
    res.is_first = rocke_b_cmp_eq(b, res.k_iter, c_zero);

    c_last = rocke_b_const_i32(b, (int64_t)(spec->k_iters - 1));
    res.is_last = rocke_b_cmp_eq(b, res.k_iter, c_last);

    return res;
}
